File: day05_elixir/part1.ex
defmodule Part1 do
    def solve do
        IO.stream(:stdio, :line)
        |> Enum.reduce(
            {%{:rules => [], :pages => []}, :mode1},
            fn line, {problem, mode} -> parse(line, problem, mode) end
        )
        |> elem(0)
        |> solve_problem()
    end

    defp parse("\n", problem, :mode1) do
        {problem, :mode2}
    end

    defp parse(line, problem, :mode1) do
        [a, b] = String.split(line, "|")
        a = String.to_integer(String.trim(a))
        b = String.to_integer(String.trim(b))
        problem = %{problem | rules: problem.rules ++ [{a, b}]}
        {problem, :mode1}
    end

    defp parse(line, problem, :mode2) do
        values = String.split(line, ",")
        values = Enum.map(values, &String.to_integer(String.trim(&1)))
        problem = %{problem | pages: problem.pages ++ [values]}
        {problem, :mode2}
    end

    defp solve_problem(problem) do
        problem.pages
        |> Enum.reject(&violates_any_rule(problem.rules, &1))
        |> Enum.map(&get_middle(&1))
        |> Enum.reduce(0, fn solved, acc -> acc + solved end)
        |> (fn result -> IO.puts("The answer is: #{result}") end).()
    end

    defp violates_any_rule(rules, a, b) do
        Enum.any?(rules, fn {x, y} -> x == b and y == a end)
    end

    defp violates_any_rule(rules, pages) do
        try do
            Enum.each(Enum.with_index(pages), fn {a, ia} ->
                Enum.each(Enum.with_index(pages), fn {b, ib} ->
                    if ib > ia and violates_any_rule(rules, a, b) do
                        throw(:violates_rule)
                    end
                end)
            end)
            false
        catch
            :violates_rule -> true
        end
    end

    defp get_middle(pages) do
        Enum.at(pages, div(length(pages), 2))
    end
end

Part1.solve()

File: day05_fsharp/part1.fs
module Part1

open System
open System.IO

type Problem = { rules: (int * int) list; pages: int list list }

let parse (problem, mode) line =
    match line, mode with
    | "", "mode1" -> problem, "mode2"
    | "", _ -> problem, mode
    | line, "mode1" ->
        let parts = line.Split('|')
        let a = Int32.Parse(parts.[0].Trim())
        let b = Int32.Parse(parts.[1].Trim())
        { problem with rules = (a, b) :: problem.rules }, "mode1"
    | line, "mode2" ->
        let values = line.Split(',')
        let values2 = values |> Array.map (fun v -> Int32.Parse(v.Trim())) |> Array.toList
        { problem with pages = values2 :: problem.pages }, "mode2"
    | _ -> problem, mode

let violatesAnyRule rules pages =
    try
        pages |> List.iteri (fun ia a ->
            pages |> List.iteri (fun ib b ->
                if ib > ia && (rules |> List.exists (fun (x, y) -> x = b && y = a)) then
                    raise (Exception "violates_rule")
            )
        )
        false
    with
    | _ -> true

let getMiddle pages =
    pages |> List.item (List.length pages / 2)

let solveProblem problem =
    problem.pages
    |> List.filter (fun pages -> not (violatesAnyRule problem.rules pages))
    |> List.map getMiddle
    |> List.sum
    |> printfn "The answer is: %d"

let solve () =
    Console.In.ReadToEnd().Split('\n')
    |> Array.fold parse ({ rules = []; pages = [] }, "mode1")
    |> fst
    |> solveProblem

solve()
